operators/pkg/instance-controller/controller_test.go

```diff
@@ -12,290 +12,408 @@
-						return true
-					}
-					return false
-				}, timeout, interval).Should(BeTrue())
+				By("Setting the instance to running", func() {
+					instance.Spec.Running = true
+					Expect(k8sClient.Update(ctx, &instance)).ToNot(HaveOccurred())
+					Expect(RunReconciler()).NotTo(HaveOccurred())
+				})
+
+				By("Asserting the right exposition resources exist", func() {
+					Expect(k8sClient.Get(ctx, forge.NamespacedName(&instance), &service)).NotTo(HaveOccurred())
+					Expect(k8sClient.Get(ctx, forge.NamespacedNameWithSuffix(&instance, forge.IngressGUINameSuffix), &ingress)).NotTo(HaveOccurred())
+					Expect(k8sClient.Get(ctx, forge.NamespacedNameWithSuffix(&instance, forge.IngressMyDriveNameSuffix), &ingress)).To(BeNotFoundError())
+				})
+
+				By("Asserting the VM has been created", func() {
+					var vmi virtv1.VirtualMachineInstance
+					Expect(k8sClient.Get(ctx, forge.NamespacedName(&instance), &vmi)).NotTo(HaveOccurred())
+					vmi.Status.Phase = virtv1.Running
+					Expect(k8sClient.Update(ctx, &vmi)).NotTo(HaveOccurred())
+				})
+
+				By("Asserting the state is coherent", func() {
+					Expect(RunReconciler()).NotTo(HaveOccurred())
+					Expect(instance.Status.Phase).To(Equal(clv1alpha2.EnvironmentPhaseRunning))
+				})
+			})
+		})
+	})
+
+	Context("In case of misconfiguration", func() {
+		When("the template is missing", func() {
+			BeforeEach(func() {
+				testName = "test-missing-template"
+				createTemplate = false
+			})
+
+			It("Should fail instance reconcile", func() {
+				Expect(RunReconciler()).To(HaveOccurred())
+			})
+		})
+
+		When("the tenant is missing", func() {
+			BeforeEach(func() {
+				testName = "test-missing-tenant"
+				createTenant = false
+			})
+
+			It("Should fail instance reconcile", func() {
+				Expect(RunReconciler()).To(HaveOccurred())
 			})
 		})
 	})
 })
 
-func doesEventuallyExist(ctx context.Context, nsLookupKey types.NamespacedName, createdObject client.Object, expectedStatus gomegaTypes.GomegaMatcher, timeout, interval time.Duration) {
-	Eventually(func() bool {
-		err := k8sClient.Get(ctx, nsLookupKey, createdObject)
-		return err == nil
-	}, timeout, interval).Should(expectedStatus)
+// IsNotFoundErrorMatcher is a custom matcher to check when kubernetes resources do not exist.
+type IsNotFoundErrorMatcher struct{}
+
+func BeNotFoundError() types.GomegaMatcher {
+	return &IsNotFoundErrorMatcher{}
+}
+
+func (s *IsNotFoundErrorMatcher) Match(actual interface{}) (success bool, err error) {
+	switch e := actual.(type) {
+	case error:
+		return errors.IsNotFound(e), nil
+	default:
+		return false, fmt.Errorf("IsNotFoundErrorMatcher can match errors only")
+	}
+}
+
+func (s *IsNotFoundErrorMatcher) FailureMessage(actual interface{}) (message string) {
+	return format.Message(actual, "to match NotFoundError")
+}
+
+func (s *IsNotFoundErrorMatcher) NegatedFailureMessage(actual interface{}) (message string) {
+	return format.Message(actual, "not to match NotFoundError")
 }
```

Comment:
Thumbs-up for this. Would just move to an utility package, to make it easier to be reused